src/de.rs

```diff
@@ -229,28 +240,31 @@ impl<'de, 'a> de::Deserializer<'de> for Deserializer {
         let s = self
             .object
             .protect_send("to_s", &[])?
-            .try_convert_to::<RString>()?
-            .to_string();
-        debug!("deserialize_string: {}", s);
-        visitor.visit_string(s)
+            .try_convert_to::<RString>()?;
+        let b = s.to_vec_u8_unchecked();
+        if str::from_utf8(&b).is_ok() {
```

Comment:
I was wondering why `String::from_utf8` wasn't used here in a similar way to above?

Reply:
Because `String::from_utf8` would consume `b` (thus preventing us from calling `visit_byte_byte` if it doesn't contain valid UTF-8). Plus, `String::from_utf8` uses `str::from_utf8` (https://doc.rust-lang.org/src/alloc/string.rs.html#482) and (basically) `String::from_utf8_unchecked` internally anyway.

Reply:
Makes sense - I was meaning something like this though:
`if let Ok(s) = String::from_utf8(s.to_vec())`
I.e. you're copying the string either way

Reply:
Yes, that's what I understood.

A few lines below there is an `else` branch which calls `visitor.visit_byte_buf(b). If we used `if let Ok(s) = String::from_utf8(s.to_vec())` we would have to replace it with `visitor.visit_byte_buf(s.to_vec())`, thus copying the bytes from Ruby twice in the worst case instead of once currently.